Implement the resize tool

jeffswenson opened the pull request:
Resolves #90, #206

jeffswenson committed:
Fix unit tests broken during the resize tool implementation.

jeffswenson committed:
Delete unused file.

NickDefo commented:
The resize tool snapping behavior should be tweaked slightly to feel more responsive and smooth.  Right now, it acts like a floor function, always snapping backwards towards the last grid line that the mouse has passed.  Instead, it should snap to the nearest grid line, regardless of whether that's forwards or backwards.

This will make the tool feel much more accurate, by moving the mouse target to where the user expects it to be.  Right now there's a 50/50 chance that a user misses the line they're trying to snap to - If they are off by even one pixel in the wrong direction, they missed.  This shift would make the tool much more forgiving, by giving the user equal room to release on either side of their target.

jeffswenson committed:
Fix resize tool for hiearchy drawables with anchors.

jeffswenson committed:
Implement resize tool. Does not handle container drawables or scale well.

tlein commented on src/duckling/canvas/tools/selected-entity-tool.ts:
```diff
@@ -0,0 +1,94 @@
+import {EntityMoveTool} from './entity-move-tool';
+import {EntityResizeTool} from './resize-tool';
+
+
```
Can you leave a class comment explaining how this is used to swap out the tool used for the selected entity depending on where it was selected from.

tlein commented on src/duckling/canvas/tools/selected-entity-tool.ts:
```diff
@@ -0,0 +1,94 @@
+        return container;
+    }
+
+    onStageDown(event : CanvasMouseEvent) {
```
You should split the logic of which tool needs to be active away from onStageDown. It's more clear that onStageDown would happen to the activeTool if it is called after this if/else chain.

jeffswenson replied:
Fixed

tlein commented on src/duckling/canvas/tools/entity-move-tool.ts:
```diff
@@ -20,51 +21,65 @@ import {drawRectangle} from '../drawing';
-                    destination, 
-                    this._getSelectedEntityBox(), 
-                    {x: gridSize, y: gridSize}));
+        if (this._selectedEntityKey) {
```
Should be an early quit:

`if (!this._selectedEntityKey) {
   return;
}`

to avoid the entire function being indented

jeffswenson replied:
fixed

tlein commented on src/duckling/canvas/tools/entity-move-tool.ts:
```diff
@@ -99,15 +113,14 @@ export class EntityMoveTool extends BaseTool {
     }
 
     private _isSnapToGrid(event : CanvasMouseEvent) {
```
This is no longer used as it's in the service

jeffswenson replied:
fixed

tlein commented on src/duckling/canvas/tools/grid-snap.service.ts:
```diff
@@ -0,0 +1,31 @@
+        return vectorAdd(position, minCornerSnapDistance(position, box, this._grid));
+    }
+
+    isSnapToGrid(event : CanvasMouseEvent) {
```
I feel this needs a better name, "shouldSnapToGrid" maybe

jeffswenson replied:
fixed

tlein commented on src/duckling/canvas/tools/entity-move-tool.ts:
```diff
@@ -99,15 +113,14 @@ export class EntityMoveTool extends BaseTool {
-            this._selectionService.selection.value.selectedEntity,
-            vectorAdd(oldPosition, adjustment),
-            this._mergeKey);
+        let entity = this._entitySystemService.getEntity(this._selectedEntityKey);
```
This doesn't get called anymore. Entities can't be adjusted via the arrow keys anymore

tlein reviewed (commented):
If you try to resize an entity with inactive drawables it will throw an error

jeffswenson committed:
Fix the resize tool for inactive drawables.

jeffswenson committed:
Fix the drag anchor pivot.

jeffswenson committed:
Clean up code duplication in the bimodal-tool and the
selected-entity-tool. Fixes a bug where arrow keys were not proxied to
the selected entity tool.

jeffswenson committed:
Fix code style comments on the resize-tool pull request.

tlein commented:
The hitboxes of the arrows are being scaled but the arrows drawn representations are not scaled. Both should not be scaled

tlein commented on src/duckling/canvas/tools/multi-mode-tool.ts:
```diff
@@ -0,0 +1,61 @@
+import {BaseTool, CanvasMouseEvent, CanvasKeyEvent} from './base-tool';
+
+/**
+ * A bimodal tool manages a primary tool and a secondary tool that is activated whenever
```
Should say "multi mode tool"

jeffswenson replied:
fixed

tlein commented on src/duckling/entitysystem/services/entity-box.service.ts:
```diff
@@ -55,11 +59,35 @@ export class EntityBoxService extends BaseAttributeService<AttributeBoundingBox>
+     * @param  box Box the entity should be resized to.
+     * @param  mergeKey Pass to merge updates on the undo/redo stack.
+     */
+    setEntityBox(entity : Entity, box : Box2, mergeKey? : string) {
```
Can you be explicit that this is returning an entity since the signature for these APIs are important?

jeffswenson replied:
fixed

tlein commented on src/duckling/entitysystem/services/resize.ts:
```diff
@@ -0,0 +1,28 @@
+function _calculateSize(valueToResize : number, boxBefore : number, boxAfter : number) {
+    return valueToResize * boxAfter / boxBefore;
+}
+// position = oldPosition - sizeChange / 2 + move
```
Is this comment supposed to be clarifying something?

jeffswenson replied:
fixed

tlein commented on src/duckling/game/drawable/drawable-drawer.ts:
```diff
@@ -55,8 +55,14 @@ export function drawDrawableAttribute(entity : Entity, assetService : AssetServi
+ * @param  assetService Service containing the assets needed to render the drawable attribute.
+ * @return DisplayObject that contains the drawn DrawableAttribute
+ */
+export function drawDrawable(drawable : Drawable, assetService : AssetService, ignoreInactive : boolean = false) : DrawnConstruct {
```
Document the ignoreInactive parameter

jeffswenson replied:
fixed

jeffswenson committed:
Hit box for drag arrows now scales with zoom level of the canvas.

jeffswenson committed:
Clean up documentation on resize-tool changes.
